src/ring_theory/noetherian.lean

```diff
@@ -508,3 +511,242 @@ nat.rec_on n
+      apply_rules @ideal.exists_maximal _ _ _,
+      let pΞ : (prime_spectrum R) := ⟨Ξ, ideal.is_maximal.is_prime hΞ⟩,
+      use ({ pΞ } : multiset (prime_spectrum R)),
+      rw [not_not, multiset.singleton_eq_singleton, multiset.map_cons,
+        multiset.map_zero, multiset.prod_singleton],
+      simp only [le_top] } },
+  obtain ⟨x, y, hx, hy, h_xy⟩ : ∃ (x y : R), x ∉ M ∧ y ∉ M ∧ x * y ∈ M,
+  { rw [ideal.is_prime, not_and] at h_not_prM,
+    specialize h_not_prM h_not_topM,
+    rw not_forall at h_not_prM,
+    cases h_not_prM with x hx,
+    rw not_forall at hx,
+    cases hx with y hxy,
+    use [x, y], finish },
+  let Jx := M + span R {x},
+  let Jy := M + span R {y},
+  have this : Jx ∉ Ω ∧ Jy ∉ Ω,
+  { have incl : ∀ (z : R), M ≤ M + span R {z},
+    { intro z, from le_sup_left },
+    have belong : ∀ (z : R), z ∈ M + span R {z},
+    { intro z,
+      apply submodule.mem_sup_right,
+      apply submodule.mem_span_singleton_self z },
+    split, all_goals {by_contradiction},
+    { have abs_x : Jx = M, from h_maxM Jx h (incl x),
+      rw ← abs_x at hx, tauto },
+    { have abs_y : Jy = M, from h_maxM Jy h (incl y),
+      rw ← abs_y at hy, tauto } },
+  obtain ⟨Wx, h_Wx⟩ : ∃ (Wx : multiset (prime_spectrum R)), multiset.prod (Wx.map subtype.val) ≤ Jx,
+  { have Jx_P : ¬ P Jx, simp [*, mem_set_of_eq, not_forall] at *,
+    dsimp [P] at Jx_P, from not_forall_not.mp Jx_P },
+  obtain ⟨Wy, h_Wy⟩ : ∃ (Wy : multiset (prime_spectrum R)), multiset.prod (Wy.map subtype.val) ≤ Jy,
+  { have Jy_P : ¬ P Jy, simp [*, mem_set_of_eq, not_forall] at *,
+    dsimp [P] at Jy_P, from not_forall_not.mp Jy_P },
+  let W := Wx + Wy,
+  have h_W : multiset.prod (W.map subtype.val) ≤ M,
+  { suffices h_JJ : multiset.prod (W.map subtype.val) ≤ Jx * Jy,
+    { suffices h_JM : Jx * Jy ≤ M, from le_trans h_JJ h_JM,
+      { dsimp only *,
+        rw add_mul,
+        have this : M * (M + span R {y}) ≤ M, from ideal.mul_le_right,
+        apply sup_le this,
+        rw mul_add,
+        replace this : span R {x} * M ≤ M, from ideal.mul_le_left,
+        apply sup_le this,
+        rw [span_mul_span, singleton_mul_singleton, span_singleton_le_iff_mem],
+        exact h_xy } },
+    { rw [multiset.map_add, multiset.prod_add],
+      apply submodule.mul_le_mul h_Wx h_Wy }, },
+  have h_absM : ¬ P M,
+  { dsimp only [P], rw not_forall_not,
+    use W, exact h_W },
+  apply absurd h_PM h_absM,
+end
+
+  /-In a noetherian integral domain which is not a field, every non-zero ideal contains a non-zero
+  product of prime ideals; in a field, the whole ring is a non-zero ideal containing only 0 as product
+  or prime ideals ([samuel, § 3.3, Lemma 3])
+-/
+variables {A : Type*} [integral_domain A] [is_noetherian_ring A]
+
+
+lemma integral_nonzero_prime_product_of_nonzero (h_fA : ¬ is_field A) (I : ideal A) (h_nzI: ⊥ < I):
+  ∃ (Z : multiset (prime_spectrum A)), multiset.prod (Z.map subtype.val) ≤ I ∧
+  multiset.prod (Z.map subtype.val) ≠ 0 :=
```

Comment:
- fit the name with the conventions
 - `I` can be implicit since it should be inferrable from `h_nzI`
 - `subtype.val` -> `coe`
 - let's indent the last line to make clear it belongs under the `∃`
 - `≠ ⊥` seems to be more popular than `⊥ <` or `≠ 0` for ideals/submodules.
```suggestion
lemma exists_prime_spectrum_prod_le_and_ne_bot (h_fA : ¬ is_field A) (I : ideal A) (h_nzI: I ≠ ⊥) :
  ∃ (Z : multiset (prime_spectrum A)), multiset.prod (Z.map (coe : subtype _ → ideal A)) ≤ I ∧
    multiset.prod (Z.map (coe : subtype _ → ideal A)) ≠ ⊥ :=
```